lib/spack/spack/util/s3.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+import six.moves.urllib.parse as urllib_parse
+
+import spack
+
+
+def create_s3_session(url):
+    parsed_url = urllib_parse.urlparse(
```

Comment:
I'm hitting an issue here when I attempt to use `util/web.py` to `push_to_url()` a file.  Below is a some output illustrating the problem, but basically, it seems like sometimes the `url` argument here may have already been parsed as a url.

```
scott@beast:/data/scott/projects/spack$ export AWS_ACCESS_KEY_ID=minio
scott@beast:/data/scott/projects/spack$ export AWS_SECRET_ACCESS_KEY=minio123
scott@beast:/data/scott/projects/spack$ export AWS_ENDPOINT_URL=http://localhost:8083
scott@beast:/data/scott/projects/spack$ source share/spack/setup-env.sh 
scott@beast:/data/scott/projects/spack$ spack python
Spack version 0.12.1
Python 3.6.8, Linux x86_64
>>> import spack.util.web as web_util
>>> dir(web_util)
['ALLOWED_ARCHIVE_TYPES', 'HTMLParseError', 'HTMLParser', 'HTTP_HEADER_NAME_ALIASES', 'LinkParser', 'NoNetworkConnectionError', 'NonDaemonContext', 'NonDaemonPool', 'NonDaemonProcess', 'Request', 'SpackWebError', 'URLError', '__UNABLE_TO_VERIFY_SSL', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_iter_s3_prefix', '_list_s3_objects', '_separators', '_spider', '_spider_wrapper', '_timeout', '_urlopen', 'codecs', 'errno', 'find_versions_of_archive', 'list_url', 'mkdirp', 'multiprocessing', 'os', 'print_function', 'product', 'push_to_url', 're', 'read_from_url', 'remove_url', 's3_util', 'shutil', 'spack', 'spider', 'ssl', 'standardize_header_names', 'string_types', 'sys', 'traceback', 'tty', 'url_exists', 'url_util', 'urllib_parse', 'urlopen', 'uses_ssl', 'warn_no_ssl_cert_checking']
>>> local_url = 'file:///tmp/junk.txt'
>>> remote_url = 's3://junkbox/stuff/junk.txt'
>>> web_util.push_to_url(local_url, remote_url)
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/data/scott/projects/spack/lib/spack/spack/util/web.py", line 254, in push_to_url
    s3 = s3_util.create_s3_session(remote_url)
  File "/data/scott/projects/spack/lib/spack/spack/util/s3.py", line 17, in create_s3_session
    allow_fragments=False)
  File "/usr/lib/python3.6/urllib/parse.py", line 367, in urlparse
    url, scheme, _coerce_result = _coerce_args(url, scheme)
  File "/usr/lib/python3.6/urllib/parse.py", line 120, in _coerce_args
    raise TypeError("Cannot mix str and non-str arguments")
TypeError: Cannot mix str and non-str arguments
>>> 
now exiting InteractiveConsole...
scott@beast:/data/scott/projects/spack$ spack python
Spack version 0.12.1
Python 3.6.8, Linux x86_64
>>> import spack.util.web as web_util
>>> local_url = 'file:///tmp/junk.txt'
>>> remote_url = 's3://junkbox/stuff/junk.txt'
>>> web_util.push_to_url(local_url, remote_url)
> /data/scott/projects/spack/lib/spack/spack/util/s3.py(16)create_s3_session()
-> parsed_url = urllib_parse.urlparse(
(Pdb) l
 11  	
 12  	
 13  	def create_s3_session(url):
 14  	    import pdb
 15  	    pdb.set_trace()
 16  ->	    parsed_url = urllib_parse.urlparse(
 17  	        url,
 18  	        scheme='file',
 19  	        allow_fragments=False)
 20  	
 21  	    if parsed_url.scheme != 's3':
(Pdb) p url
ParseResult(scheme='s3', netloc='junkbox', path='/stuff/junk.txt', params='', query='', fragment='')
(Pdb)
```

Reply:
The following patch fixed it for me and I was able to push the local file into the S3 bucket:

```
scott@beast:/data/scott/projects/spack$ git diff
diff --git a/lib/spack/spack/util/s3.py b/lib/spack/spack/util/s3.py
index 2d0a34a29..095b057b9 100644
--- a/lib/spack/spack/util/s3.py
+++ b/lib/spack/spack/util/s3.py
@@ -11,10 +11,12 @@ import spack
 
 
 def create_s3_session(url):
-    parsed_url = urllib_parse.urlparse(
-        url,
-        scheme='file',
-        allow_fragments=False)
+    parsed_url = url
+    if not isinstance(parsed_url, urllib_parse.ParseResult):
+        parsed_url = urllib_parse.urlparse(
+            url,
+            scheme='file',
+            allow_fragments=False)
 
     if parsed_url.scheme != 's3':
         raise ValueError(
scott@beast:/data/scott/projects/spack$
```

Reply:
To be clear, I'm not suggesting that is necessarily the right solution, but hopefully it makes clear what could be going wrong.